src/operations.c

```diff
@@ -141,32 +164,25 @@ static int iterate_json(FILE *input, rodsEnv *env, baton_json_op fn,
         (*item_count)++;
 
         json_decref(item); // JSON free
-
-        time_t now = time(0);
-        double duration = difftime(now, connect_time);
-        if (args->max_connect_time > 0 && duration > args->max_connect_time) {
-            logmsg(INFO, "The connection to iRODS was open for %d seconds, "
-                   "the maximum allowed is %d; closing the connection to "
-                   "reopen a new one", duration, args->max_connect_time);
-            rcDisconnect(conn);
-            conn      = NULL;
-            reconnect = 1;
-        }
     } // while
 
     if (exit_flag) {
       status = exit_flag;
       logmsg(WARN, "Exiting on signal with code %d", exit_flag);
       goto finally;
     }
-    
-    if (drop_conn_count > 0) {
-      logmsg(WARN, "Reconnected for put operations %d times",
-             drop_conn_count);
-    }
 
 finally:
-    if (conn) rcDisconnect(conn);
+    pthread_mutex_lock(&conn_mutex);
+    run_timeout_thread = 0;
+    pthread_cond_signal(&watchdog_cond); // Unblock the thread waiting on cond
+
+    if (connection) {
+        rcDisconnect(connection);
+        connection = NULL;
+        logmsg(NOTICE, "Closed the connection on exit")
+    }
+    pthread_mutex_unlock(&conn_mutex);
```

Comment:
It probably doesn't matter too much as your program is about to exit, but you should probably do a `pthread_join()` here to clean up the thread you made above (if `pthread_create()` worked).  Not doing so leaves a zombie thread behind, albeit not for long if you exit soon after.  It would be important if you ever want to call this function more than once in a program invocation.